# Sprint 3

## Ejercicio 1

1. Crea una funcion que te ha de dar un Array de una dimensión, que además haga un resumen estadístico básico de los datos.


2. Si detecta que el Array tiene más de una dimensión, ha de mostrar un mensaje de error.


Primero se importa la librería Numpy y el módulo Random:

In [8]:
import numpy as np
from numpy import random

Se crea una función que dará como resultado un array de una dimensión.  Para lograrlo, se usa el método `randint()` del módulo `Random` con números entre 1 y 100. 

In [9]:
def array_1D(n):
    x = np.random.randint(100, size=n)
    return x

Después se crea una segunda función a la que se aplicarán las funciones de _máximo, mínimo, media y variación estándar_.

Para poder detectar si es un array de más de una dimensión, se agregó la condición `if` en la que se establece si el número de dimensiones no es igual a 1 `(...ndim != 1)`entonces regresara la leyenda **"Error: No es un array de una dimensión".**

Finalmente, se pide que se muestren los números aleatorios del `array` con las leyendas de `Maximo, Minimo, Media y Variación Estandar` para tener mayor claridad en los resultados.

In [10]:
def resum_estadistico(array):
    max_valor = np.max(array)
    min_valor = np.min(array)
    media_valor = np.mean(array)
    var_stand_valor = np.std(array)
    if array.ndim != 1:
        return 'Error: No es un array de una dimensión'
    else:
        return array, 'Maximo ' + str(max_valor), 'Minimo ' + str(min_valor), 'Media ' + str(media_valor), 'Variació Estandar ' + str(var_stand_valor)
    


Para llamar a ambas funciones, se crea una variable nueva que contiene a la función `array_1D()` en el que se especifica el número de valores que se desean para después llamar a la función `resum_estadistico()`.

In [11]:
resultado = array_1D(5)
resum_estadistico(resultado)

(array([88, 67, 30, 52, 58]),
 'Maximo 88',
 'Minimo 30',
 'Media 59.0',
 'Variació Estandar 18.952572384771415')

Para que la salida muestre un error, se debe crear un array de más de una dimensión y después llamar a la función `resum_estdistico()`

In [12]:
array_2D = np.array([[1,2,3],[4,5,6]])
resum_estadistico(array_2D)

'Error: No es un array de una dimensión'

## Ejercicio 2

1. Crea una función que te genere un cuadrante NXN de números aleatorios entre el 0 y el 100.

Primero se importa el módulo `randint()` desde Random.

In [17]:
from random import randint

Después se crea una función que integra dos bucles `for` para iterar los elementos dentro de un rango `(range())` de la variante `fila = []` en el primer bucle y en el segundo usando el método `append()`, se da la instrucción de `randint()` que toma números aleatorios de 1 al 100.

In [9]:
def cuadrante_NN(n):
    matriz = []
    for f in range(n):
        fila = []
        for c in range(n):
            fila.append(randint(1,100))
        matriz.append(fila)
    return matriz

Para llamar la funcion `cuadrante_NN()` se especifica el número de arrays deseado.

In [21]:
cuadrante_NN(5) #output: 5 filas y 5 columnas

[[57, 38, 73, 75, 43],
 [94, 6, 68, 45, 10],
 [35, 37, 32, 15, 31],
 [15, 33, 64, 28, 86],
 [41, 28, 33, 16, 32]]

## Ejercicio 3

1. Crea una función que dada una tabla de dos dimensiones (NxM), calcule los totales por fila y los totales por columna.

Primero se llama a la librería de Numpy. 

In [22]:
import numpy as np #este paso es solo para ejemplificar, no es necesario si ya se llamó antes a la librería.

Se crea una función `crear_tabla()` que contiene dos variables con la función `np.sum()` para sumar los números dentro de ese array.  La distinción entre fila y columna, se hace al especificar el `axis=` - filas es _axis=1_ y columnas es _axis=0_.

In [23]:
def crear_tabla(tabla):
    sum_filas= np.sum(tabla, axis=1)
    sum_colum = np.sum(tabla, axis=0)
    return 'Suma de filas: ' + str(sum_filas), 'Suma de columnas' + str(sum_colum)

Para llamar esta función en una tabla dada de dos dimensiones (NxM), se puede hacer de dos maneras como se muestra a continuación.
    
Primero sólo creando el array y llamando a la función:

In [24]:
array_2D = np.array([[1,2,3,4],[5,6,7,8]])
crear_tabla(array_2D) #este output se ve demasiado junto y desordenado.

('Suma de filas: [10 26]', 'Suma de columnas[ 6  8 10 12]')

Segundo, se crea el mismo array y después una variable que contenga todo: `sum_filas, sum_colum`, la función `crear_tabla y array_2D` para una impresión más limpia y separada:

In [25]:
array_2D = np.array([[1,2,3,4],[5,6,7,8]])
sum_filas, sum_colum = crear_tabla(array_2D)
print(sum_filas)
print(sum_colum)

Suma de filas: [10 26]
Suma de columnas[ 6  8 10 12]


## Ejercicio 4

1. Implementa manualmente una función que calcule el coeficiente de correlación.
2. Informate sobre sus usos e interpretación.

**"El coeficiente de correlación hace algo aparentemente milagroso: colapsa un complejo conjunto de datos medidos en diferentes unidades (como altura y peso) en una sola y elegante estadística descriptiva."** (Charles Wheelan, Naked Statistics, 2013)

Primero se deben hacer los siguientes calculos sobre las variables `altura` y `peso`.

In [33]:
#se crean las dos variables
altura = [1.70, 1.65, 1.60, 1.55, 1.69]
peso = [80, 63, 90, 65, 100]

#se calcula la media de altura y de peso
altura_media = np.mean(altura)
peso_media = np.mean(peso)

#se calcula la desviación estandar de ambos
altura_std = np.std(altura)
peso_std = np.std(peso)

print('Media altura: ', altura_media)
print('Desviación estandar altura: ', altura_std)
print('Media peso: ', peso_media)
print('Desviación estandaar peso: ', peso_std)

Media altura:  1.638
Desviación estandar altura:  0.05635601121442143
Media peso:  79.6
Desviación estandaar peso:  14.235167719419396


Después, para calcular el coeficiente de correlación, se deben realizar los siguientes pasos que a su vez se integrarán en una función:

1. Convertir cada valor de la variable `altura` en unidades estandar: (altura - media / variación estandar)
2. Convertir cada valor de la variable `peso` en unidades estandar: (peso - media / variación estandar)
3. Calcular el producto de cada valor de las variables de (`peso en unidades estandar`) x (`altura en unidades estanda`)
4. El *coeficiente de correlación* es la suma de los productos calculados anteriormente, dividido por el número de observaciones

Para lograrlo, se crea una variable para `altura` y otra para `peso` que incluya un bucle `for` que itinere cada valor dentro de esa variable `len`.

In [40]:
altura_unidad_estandar = [(altura[i] - altura_media) / altura_std for i in range(len(altura))]
peso_unidad_estandar = [(peso[i] - peso_media) / peso_std for i in range(len(peso))]

print('Unidad Estandar de Altura: ', altura_unidad_estandar)
print('Unidad Estandar de Peso: ', peso_unidad_estandar)

Unidad Estandar de Altura:  [1.1001488335309, 0.21293203229630323, -0.6742847689382896, -1.5615015701728865, 0.9227054732839807]
Unidad Estandar de Peso:  [0.028099423054519537, -1.1661260567625438, 0.730584999417498, -1.0256289414899482, 1.4330705757804765]


Al ser los resultados demasiado grandes, se redondean usando la función `round()` en cada una de las variables y especificando que sea a 3 decimales.

In [46]:
altura_unidad_estandar = [round(x, 3) for x in altura_unidad_estandar]
peso_unidad_estandar = [round(x, 3) for x in peso_unidad_estandar]

print('Unidad Estandar de Altura: ', altura_unidad_estandar)
print('Unidad Estandar de Peso: ', peso_unidad_estandar)

Unidad Estandar de Altura:  [1.1, 0.213, -0.674, -1.562, 0.923]
Unidad Estandar de Peso:  [0.028, -1.166, 0.731, -1.026, 1.433]


Para el paso 3, se multiplican los valores de las variables de `unidad estandar`. Para lograrlo se usa `[i]` para acceder a cada uno de los valores dentro de la variable, al mismo tiempo que se iteran con el bucle `for`.  Esta vez hecho, se redondea desde esta variable.

In [47]:
producto_altura_peso = [round(peso_unidad_estandar[i] * altura_unidad_estandar[i], 3) for i in range(len(altura))]

print('Producto de altura y peso: ', producto_altura_peso)

Product de altura y peso:  [0.031, -0.248, -0.493, 1.603, 1.323]


Para el paso 4, primero se suman los productos obtenidos con la función `np.sum()` de Numpy.

In [48]:
suma_productos = np.sum(producto_altura_peso)

print(suma_productos)

2.216


Y para terminar, se divide el resultado de la suma entre el número de elementos dentro de las variables originales de `peso` y `altura` para después crear una variable que redondeé todo a tres decimales.

In [52]:
coeficiente_correlacion = suma_productos/5
coeficiente_correlacion_redondeado = round(coeficiente_correlacion, 3)

print('Coeficiente de Correlación: ',coeficiente_correlacion_redondeado)

Coeficiente de Correlación:  0.443


Después de haber hecho todos los calculos manualmente, se unen en la función `coef_corr`que dará como resultado el _coeficiente de correlación_.

In [1]:
def coef_corr (altura,peso):
    altura_media = np.mean(altura)
    peso_media = np.mean(peso)
    altura_std = np.std(altura)
    peso_std = np.std(peso)
    altura_unidad_estandar = [(altura[i] - altura_media) / altura_std for i in range(len(altura))]
    peso_unidad_estandar = [(peso[i] - peso_media) / peso_std for i in range(len(peso))]
    altura_unidad_estandar = [round(x, 3) for x in altura_unidad_estandar]
    peso_unidad_estandar = [round(x, 3) for x in peso_unidad_estandar]
    producto_altura_peso = [round(peso_unidad_estandar[i] * altura_unidad_estandar[i], 3) for i in range(len(altura))]
    suma_productos = np.sum(producto_altura_peso)
    coeficiente_correlacion = suma_productos/len(altura)
    coeficiente_correlacion_redondeado = round(coeficiente_correlacion, 3)
    return coeficiente_correlacion_redondeado
    

Se llama a la función `coef_corr` con los datos de las variables originales:

In [58]:
altura = [1.70, 1.65, 1.60, 1.55, 1.69]
peso = [80, 63, 90, 65, 100]

coef_corr(altura, peso)

0.443

El coeficiente de correlación de 0.443 indica que hay una relación **positiva moderada** entre el `peso` y la `altura`. Esto significa que hay una tendencia general de que las personas más altas tienden a pesar más.

Sin embargo, sin más contexto que saber cuál es el coeficiente de correlación entre estas variantes, no se puede puede concluir que entre más alta sea una persona, tendrá mayor peso; o que la altura es la causa de tener mayor peso; o viceversa.  Esto es porque hay varios factores externos que influyen, como la dieta y las costumbres alimenticias.

# FIN